typing/mtype.ml

```diff
@@ -71,40 +79,56 @@ and strengthen_sig ~aliasable env sg p =
             else
               { decl with type_manifest = manif }
       in
-      Sig_type(id, newdecl, rs, vis) :: strengthen_sig ~aliasable env rem p
-  | (Sig_typext _ as sigelt) :: rem ->
-      sigelt :: strengthen_sig ~aliasable env rem p
-  | Sig_module(id, pres, md, rs, vis) :: rem ->
+      SigL_type(id, newdecl, rs, vis) :: strengthen_lazy_sig' ~aliasable env rem p
+  | (SigL_typext _ as sigelt) :: rem ->
+      sigelt :: strengthen_lazy_sig' ~aliasable env rem p
+  | SigL_module(id, pres, md, rs, vis) :: rem ->
       let str =
-        strengthen_decl ~aliasable env md (Pdot(p, Ident.name id))
+        strengthen_lazy_decl ~aliasable env md (Pdot(p, Ident.name id))
       in
-      Sig_module(id, pres, str, rs, vis)
-      :: strengthen_sig ~aliasable
-        (Env.add_module_declaration ~check:false id pres md env) rem p
+      let env =
+        Env.add_module_declaration_lazy ~update_summary:false id pres md env in
+      SigL_module(id, pres, str, rs, vis)
+      :: strengthen_lazy_sig' ~aliasable env rem p
       (* Need to add the module in case it defines manifest module types *)
-  | Sig_modtype(id, decl, vis) :: rem ->
+  | SigL_modtype(id, decl, vis) :: rem ->
       let newdecl =
-        match decl.mtd_type with
-          None ->
-            {decl with mtd_type = Some(Mty_ident(Pdot(p,Ident.name id)))}
-        | Some _ ->
+        match decl.mtdl_type with
+        | Some _ when not aliasable ->
```

Comment:
We don't need this case. `aliasable` tells you whether the path is one that can safely be used as module alias -- i.e. no applications and no functor parameters -- but there are no such restrictions on module type aliases.

We could maybe leave the existing module type if it is of the form `MtyL_ident`. Just to do fewer expansions when accessing it later, but I'm not sure it is worth it.

Reply:
I thought so too! We do need it because of a tricky case with recursive modules. See fstclassmod.ml in the testsuite. (I'll add a comment)